### PR TITLE
[FSDP] CPU offload resubmit

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #67249

Implements CPU offload for model parameters in FSDP. Rehash of https://github.com/pytorch/pytorch/pull/66959.

- CPU offload class with only offload_params attribute is created
- If this is specified in FSDP ctor, model parameters are moved back to CPU after sharding in __init__
- In forward pass, during lazy init, p._local_shard gets set to p.data so it is on CPU. We pin_memory here.
- In forward pass, in _rebuild_full_params, we move p.data back to self.compute_device if necessary. Note that we don't use the device of p._full_param_padded because we don't always have this attr, but when we do its always the same as compute_device.
- The same logic as above applies to the beginning of backwards pass.
- At end of fwd and end of bwd, `_use_param_local_shard` takes care to ensure the parameters are offloaded to CPU again, by pointing it to p._local_shard, which is always on CPU.
- Note that gradient is also offloaded in _post_backward_hook. This is so model param and grad can reside on the same device.

Regarding tests:
- We tests 3 different types of init: 1) CUDA the model before wrapping with FSDP, 2) CUDA the model after wrapping with FSDP, 3) never CUDA the model.
- Case 1 is always supported. Case 2 is not supported with CPU offload and throws an error during fwd pass. Case 3 is only supported with CPU offload at the moment.
- Verifies all params are offloaded to CPU after init.
- Verifies all params are offloaded to CPU after forward and backward.
- Verifies parity with DDP. 

Differential Revision: [D31911085](https://our.internmc.facebook.com/intern/diff/D31911085/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

### CODE REVIEW

**Review by zhaojuanmao — COMMENTED** [review]
nice test cases and refactoring! as we discussed offline, let us add moving grads back to CPU as well, also we can use parametrizing tests to simplify tests

**Review by zhaojuanmao — COMMENTED** [review]
overall looks good! would you please change to non-blocking copy grads back to CPU as well?

**Review by zhaojuanmao — APPROVED** [review]
approving to unblock, it is still good to run benchmarks and check the results and streams in the traces before landing though

File: torch/distributed/_fsdp/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -343,10 +379,33 @@ def _init_param_attributes(self, p: Parameter) -> None:
             p, "_orig_size"
         ), "Parameters should have been sharded during construction."
         if hasattr(p, "_local_shard"):
+            # If CPU offloading, p._local_shard should have been placed on CPU
+            # during its first lazy construction.
+            if self.cpu_offload.offload_params:
+                assert (
+                    p._local_shard.device == torch.device("cpu")  # type: ignore[attr-defined]
+                ), "Expected p._local_shard to be on CPU."  # type: ignore[attr-defined]
```
Comment (mrshenli) [inline torch/distributed/_fsdp/fully_sharded_data_parallel.py:387]: nit: maybe add a ", but it's on {p._local_shard.device}"?

File: torch/distributed/_fsdp/fully_sharded_data_parallel.py [hunk 2]
```diff
@@ -343,10 +379,33 @@ def _init_param_attributes(self, p: Parameter) -> None:
...
             return
 
-        # A single shard of the parameters.
+        # A single shard of the parameters. Also makes p._local_shard to be on
+        # CPU if we are CPU offloading, since p.data would be on CPU during
+        # init.
+        if self.cpu_offload.offload_params:
```
Comment (mrshenli) [inline torch/distributed/_fsdp/fully_sharded_data_parallel.py:393]: would I be correct that the implementation is the same as the following?

```python
if self.cpu_offload.offload_params:
  if hasattr(p, "_local_shard"):
    assert (
      p._local_shard.device == torch.device("cpu")  # type: ignore[attr-defined]
    ), "Expected p._local_shard to be on CPU."  # type: ignore[attr-defined]
    return

  assert p.device == torch.device("cpu"), "Expected param to be on CPU when cpu_offloading is enabled."
  p._local_shard = p.data  # type: ignore[attr-defined]
  p._local_shard.pin_memory()  # type: ignore[attr-defined]
  p._cpu_grad = torch.zeros_like(  # type: ignore[attr-defined]
    ...
  ).pin_memor()
else:
  p._local_shard = p.data  # type: ignore[attr-defined]
  

  
```

**Review by mrshenli — COMMENTED** [review]
Will come back to the rest of this shortly.

File: torch/distributed/_fsdp/fully_sharded_data_parallel.py [hunk 3]
```diff
@@ -651,6 +730,13 @@ def _wait_for_post_backward(self) -> None:
             self._assert_state(TrainingState_.BACKWARD_PRE)
 
         torch.cuda.current_stream().wait_stream(self._streams["post_backward"])
+        if self.cpu_offload.offload_params:
+            # We need to wait for the non-blocking GPU ->
+            # CPU grad transfers to finish. TODO investigate why this is needed
```
Comment (mrshenli) [inline torch/distributed/_fsdp/fully_sharded_data_parallel.py:735]: I think this is because, when the grad is on CPU, CPU comp won't wait for any CUDA stream (including default stream), unless we explicitly block CPU thread using `torch.cuda.current_stream().synchronize()`.

File: torch/testing/_internal/common_fsdp.py [hunk 4]
```diff
@@ -17,7 +20,20 @@
     get_cycles_per_ms,
 )
 
-# get full params of a model recursively
+from enum import Enum
```
Comment (mrshenli) [inline torch/testing/_internal/common_fsdp.py:23]: nit: move this up to put it together with other non-torch imports?

**Review by mrshenli — APPROVED** [review]
FSDP source code changes LGTM! Didn't read test code. All my comments are minor. Please feel free to address in followup PRs.

File: torch/distributed/_fsdp/fully_sharded_data_parallel.py (hunk 3 above)
Comment (rohan-varma) [reply torch/distributed/_fsdp/fully_sharded_data_parallel.py:735]: Ah, that makes sense. I'll update the comment here to avoid confusion later.

File: torch/distributed/_fsdp/fully_sharded_data_parallel.py (hunk 2 above)
Comment (rohan-varma) [reply torch/distributed/_fsdp/fully_sharded_data_parallel.py:393]: Yep, although in the above code, the part before and until `return` is redundant right? We check that above after L381? I feel like the current code is a bit more clear that `p._local_shard = p.data` always regardless of CPU offload or not, but open to changing it to this as well depending on what you think.

File: torch/distributed/_fsdp/fully_sharded_data_parallel.py [hunk 5]
```diff
@@ -343,10 +379,33 @@ def _init_param_attributes(self, p: Parameter) -> None:
...
+        # CPU if we are CPU offloading, since p.data would be on CPU during
+        # init.
+        if self.cpu_offload.offload_params:
+            assert p.device == torch.device("cpu"), "Expected param to be on CPU when cpu_offloading is enabled."
         p._local_shard = p.data  # type: ignore[attr-defined]
+        # If CPU offloading, pin the memory to enable faster CPU -> GPU device
+        # transfer.
+        if self.cpu_offload.offload_params:
+            assert p._local_shard.device == torch.device("cpu")  # type: ignore[attr-defined]
+            p._local_shard.pin_memory()  # type: ignore[attr-defined]
```
Comment (rohan-varma) [inline torch/distributed/_fsdp/fully_sharded_data_parallel.py:400]: I don't think this actually pins the memory, it needs to be `p = p.pin_memory()`
